src/session-pool.js

```diff
@@ -435,252 +366,370 @@ SessionPool.prototype.requestStream = function(config) {
+SessionPool.prototype.listenForEvents_ = function() {
+  var self = this;
+
+  var keepAliveInterval = this.options.keepAlive * 60000;
+  var evictInterval = this.options.idlesAfter * 60000;
+
+  this.once('available', startIntervals);
+
+  this.once('close', function() {
+    killIntervals();
+
+    self.removeListener('available', startIntervals);
+    self.removeListener('destroy', onDestroy);
   });
 
-  if (this.acquireIntervalId) {
-    return;
-  }
+  function startIntervals() {
+    self.on('destroy', onDestroy);
 
-  var self = this;
-  var intervalSpeed = 30000;
+    self.pingHandle_ = setInterval(function() {
+      self.pingIdleSessions_();
+    }, keepAliveInterval);
 
-  this.acquireIntervalId = setInterval(checkForSession, intervalSpeed);
+    self.evictHandle_ = setInterval(function() {
+      self.evictIdleSessions_();
+    }, evictInterval);
+  }
 
-  function checkForSession() {
-    var hasFreeSession =
-      self.pool.free || (self.writePool && self.writePool.free);
+  function onDestroy() {
+    if (!self.available_.length) {
+      killIntervals();
 
-    if (hasFreeSession) {
-      var acquireReq = self.pendingAcquires.shift();
-      self.getNextAvailableSession_(acquireReq.options, acquireReq.callback);
+      self.removeListener('destroy', onDestroy);
+      self.once('available', startIntervals);
     }
+  }
 
-    if (!self.pendingAcquires.length) {
-      clearInterval(self.acquireIntervalId);
-      self.acquireIntervalId = null;
-    } else if (self.acquireTimeout) {
-      var err = new Error('Unable to acquire Session, timeout occurred.');
-      var acquire;
-
-      for (var i = self.pendingAcquires.length - 1; i > -1; i--) {
-        acquire = self.pendingAcquires[i];
-        acquire.timeout -= intervalSpeed;
-
-        if (acquire.timeout < 0) {
-          self.pendingAcquires.splice(i, 1);
-          acquire.callback(err);
-        }
-      }
-    }
+  function killIntervals() {
+    clearInterval(self.pingHandle_);
+    clearInterval(self.evictHandle_);
   }
 };
 
 /**
- * Release a write session. This will also begin a new transaction before
- * returning it to the pool.
+ * Makes a keep alive request to all the idle sessions.
  *
  * @private
  *
- * @param {Session} session - The session to be released.
+ * @fires SessionPool#error
  * @returns {Promise}
  */
-SessionPool.prototype.releaseWriteSession_ = function(session) {
+SessionPool.prototype.pingIdleSessions_ = function() {
   var self = this;
-  var transaction = self.createTransaction_(session);
 
-  return transaction
-    .begin()
-    .then(function() {
-      session.transaction_ = transaction;
-      return self.writePool.release(session);
-    })
-    .then(null, function() {
-      return self.writePool.destroy(session);
-    });
+  var sessions = this.getIdleSessions_();
+  var requests = sessions.map(function(session) {
+    return self.requestQueue_
+      .add(function() {
+        return session.keepAlive();
```

Comment:
we should rest session.txn when this returns. That transaction is no longer valid because for doing keep alive a new implicit transaction was started.

Reply:
This is still pending.